File: 1160-letter-tile-possibilities/letter-tile-possibilities.py
class Solution:
    def numTilePossibilities(self, tiles: str) -> int:
        res = set()
        def dfs(part,tile):
            if(part):
                res.add(part)
            for i in range(len(tile)):
                dfs(part+tile[i],tile[:i]+tile[i+1:])
        dfs('',tiles)
        return len(res)

File: 1160-letter-tile-possibilities/letter-tile-possibilities.cpp
class Solution {
public:
    unordered_set<string> res;
    void dfs(string part, string tile) {
        if (!part.empty()) {
            res.insert(part);
        }
        for (int i = 0; i < tile.size(); i++) {
            dfs(part + tile[i],
                tile.substr(0, i) + tile.substr(i + 1));
        }
    }
    int numTilePossibilities(string tiles) {
        dfs("", tiles);
        return res.size();
    }
};
